HerPortal.BusinessLogic/ExternalServices/S3FileReader/S3FileReader.cs

```diff
@@ -0,0 +1,103 @@
+﻿using Amazon;
+using Amazon.S3;
+using Amazon.S3.Model;
+using Amazon.S3.Transfer;
+using HerPublicWebsite.BusinessLogic.ExternalServices.S3FileWriter;
+using HerPublicWebsite.BusinessLogic.Services.S3ReferralFileKeyGenerator;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace HerPortal.Data.ExternalServices.S3FileReader;
+
+public class S3FileReader : IS3FileReader
+{
+    private readonly S3FileReaderConfiguration config;
+    private readonly S3ReferralFileKeyService keyService;
+
+    private readonly ILogger logger;
+
+    public S3FileReader
+    (
+        IOptions<S3FileReaderConfiguration> options,
+        S3ReferralFileKeyService keyService,
+        ILogger<S3FileReader> logger
+    ) {
+        this.config = options.Value;
+        this.keyService = keyService;
+
+        this.logger = logger;
+    }
+    
+    public async Task<Stream> ReadFileAsync(string custodianCode, int year, int month)
+    {
+        var key = keyService.GetS3KeyFromData(custodianCode, year, month);
+
+        try
+        {
+            var s3Client = new AmazonS3Client(RegionEndpoint.GetBySystemName(config.Region));
```

Comment:
I think the S3 client is designed to be reused, so this could be initialised in the constructor, if you prefer.